qadence/analog/addressing.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from warnings import warn
+
+import sympy
+import torch
+from numpy import pi
+
+from qadence.parameters import Parameter, evaluate
+from qadence.types import StrEnum
+
+DEFAULT_MAX_AMPLITUDE = 2 * pi * 3
+DEFAULT_MAX_DETUNING = 2 * pi * 20
+
+
+class WeightConstraint(StrEnum):
+    """Supported types of constraints for addressing weights."""
+
+    NORMALIZE = "normalize"
+    """Normalize weights so that they sum up to 1."""
+
+    RESTRICT = "restrict"
+    """Restrict weight values to interval [0, 1]."""
+
+
+def sigmoid(x: torch.Tensor, a: float, b: float) -> sympy.Expr:
+    return 1.0 / (1.0 + sympy.exp(-a * (x + b)))
+
+
+@dataclass
+class AddressingPattern:
+    # number of qubits
+    n_qubits: int
+
+    # list of weights for fixed amplitude pattern that cannot be changed during the execution
+    weights_amp: dict[int, float | torch.Tensor | Parameter]
+
+    # list of weights for fixed detuning pattern that cannot be changed during the execution
+    weights_det: dict[int, float | torch.Tensor | Parameter]
+
+    # maximum amplitude can also be chosen as a variational parameter if needed
+    max_amp: float | torch.Tensor | Parameter = DEFAULT_MAX_AMPLITUDE
+
+    # maximum detuning can also be chosen as a variational parameter if needed
+    max_det: float | torch.Tensor | Parameter = DEFAULT_MAX_DETUNING
```

Comment:
There might be something I am missing, but won't it be easier to maintain if you only had `weights` and `max_value` instead of `weights_amp`/`weights_det` and `max_amp`/`max_det` ? It seems to me that you are doing the same operations on `weights_amp` and `weights_det`, and on `max_amp` and `max_det`. You would either have to generate two instances of AddressingPattern (one for amp and one for det) or create a class that takes weights_amp,weights_det and max_amp, max_det and uses the modified AddressingPattern to update values ?

Reply:
I see now your point. From code perspective the applied transformations for amplitude/detuning weights are similar, however how they integrate in later stages of `qadence` execution pipeline is different. So I'm currently refactoring the duplicated code into functions but keeping separate `weights_amp`, `weights_det`, etc.